### PR TITLE
[2942] Add the Card DnD capability on Deck

### LINKED ISSUE #2942: Add the Card Drag and Drop support
* [x] **I have checked that this feature has not yet been suggested by someone else.**

### Expected behavior

It is possible to define a Drop Tool on a Lane Description to specify the behavior in the model when a Card is dropped on this Lane.
When a card is dragged and dropped within or toward an other lane, the drop tool described in the deck view model is executed.
...

### Cuttings backs

...

### Potential side effects

...

### Describe alternatives you've considered

...

### PR DESCRIPTION
**Pull request by florianbarbin**
- Make it possible to specify the card drop behavior on a Lane in the Deck view model.
- Add the DropCard mutation to handle the front-end drag and drop and execute the tool.

Bug: https://github.com/eclipse-sirius/sirius-web/issues/2942

# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [x] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?

- Create a new Blank Studio and create a Deck view by using the deck stereotype
- Create a new Task project from the template
- Create a new Deck Representation on the Daily Dev Project
- Try to DnD a card within a Lane or to an other lane: The ordering should be persisted and reflected in the model

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: packages/starters/backend/sirius-components-task-starter/src/main/java/org/eclipse/sirius/components/task/starter/configuration/view/TaskJavaService.java [hunk 1]
```diff
@@ -145,4 +144,52 @@ public void editCard(EObject eObject, String title, String description, String l
...
+     *            the task to move.
+     * @param index
+     *            the new index in the lane task list.
+     * @param targetLaneTaskList
+     *            the current lane task list.
+     * @param project
+     *            the project owning the tasks.
+     * @return the index on which the task should be moved in the project task list to match the new index in the lane.
+     */
+    private int computeIndexOfaskToReplace(Task task, int index, List<Task> targetLaneTaskList, Project project) {
```
Comment (frouene) [inline packages/starters/backend/sirius-components-task-starter/src/main/java/org/eclipse/sirius/components/task/starter/configuration/view/TaskJavaService.java]: typo in the method name ask -> task

File: packages/deck/backend/sirius-components-deck/src/main/java/org/eclipse/sirius/components/deck/description/LaneDescription.java [hunk 2]
```diff
@@ -29,7 +29,17 @@
...
 public record LaneDescription(String id, Function<VariableManager, String> targetObjectKindProvider, Function<VariableManager, String> targetObjectLabelProvider,
         Function<VariableManager, String> targetObjectIdProvider, Function<VariableManager, List<Object>> semanticElementsProvider, Function<VariableManager, String> titleProvider,
-        Function<VariableManager, String> labelProvider, List<CardDescription> cardDescriptions, Consumer<VariableManager> editLaneProvider, Consumer<VariableManager> createCardProvider) {
+        Function<VariableManager, String> labelProvider, List<CardDescription> cardDescriptions, Consumer<VariableManager> editLaneProvider, Consumer<VariableManager> createCardProvider, Consumer<VariableManager> dropCardProvider) {
+
+
+    public static final String NEW_LABEL = "newLabel";
+    public static final String TITLE = "title";
+    public static final String DESCRIPTION = "description";
+    public static final String OLD_LANE = "oldLane";
```
Comment (frouene) [inline packages/deck/backend/sirius-components-deck/src/main/java/org/eclipse/sirius/components/deck/description/LaneDescription.java]: The first three variables are never used

**Review by frouene — CHANGES_REQUESTED** [review]

File: packages/starters/backend/sirius-components-task-starter/src/main/java/org/eclipse/sirius/components/task/starter/configuration/view/TaskJavaService.java (hunk 1 above)
Comment (florianbarbin) [reply packages/starters/backend/sirius-components-task-starter/src/main/java/org/eclipse/sirius/components/task/starter/configuration/view/TaskJavaService.java]: done

File: packages/deck/backend/sirius-components-deck/src/main/java/org/eclipse/sirius/components/deck/description/LaneDescription.java (hunk 2 above)
Comment (florianbarbin) [reply packages/deck/backend/sirius-components-deck/src/main/java/org/eclipse/sirius/components/deck/description/LaneDescription.java]: done

**Review by frouene — APPROVED** [review]

**Pull request merged**